### PR TITLE
Add rules for revert reason string

### LINKED ISSUE #108: Add rules for revert reason strings
We'll be working on [revert reasons in OpenZeppelin](https://github.com/OpenZeppelin/openzeppelin-solidity/issues/1709), and it'd be great to have some sort of linter support for this. The two features I can think of are:
a) check that every `require` statement has a reason string
b) check that each reason string is at most N characters long (if it's a literal)

Do you think this is something that makes sense to have in solhint?

### PR DESCRIPTION
**Pull request by mariano-aguero**
Closes #108 

This is an approach to add rules for revert/require reason strings.
Take into account the following:
- Check that every require or revert statement has a reason string
- Check that each reason string is at most N characters long (if it's a literal)

Regards

### CODE REVIEW

File: test/rules/best-practises/reason-string.js [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+    assertErrorMessage(report, 'Error message for revert is too long')
+  })
+
+  it('should not raise warning for require', () => {
+    const code = funcWith(`require(!has(role, account), "Roles: account already has role");
+        role.bearer[account] = true;role.bearer[account] = true;
+    `)
+
+    const report = linter.processStr(code, {
+      rules: { 'reason-string': ['warn', { maxLength: 50 }] }
```
Comment (fvictorio) [inline test/rules/best-practises/reason-string.js]: Make `maxLength` 31 (the exact length of the reason string in the snippet) to make sure there aren't off-by-one errors.

File: test/rules/best-practises/reason-string.js [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+
+    const report = linter.processStr(code, {
+      rules: { 'reason-string': ['error', { maxLength: 50 }] }
+    })
+
+    assertNoWarnings(report)
+    assertNoErrors(report)
+  })
+
+  it('should not raise warning for require with text errors', () => {
```
Comment (fvictorio) [inline test/rules/best-practises/reason-string.js]: I don't understand what is being tested in these last two scenarios.

File: lib/rules/best-practises/reason-string.js [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+  defaultSetup: ['warn', { maxLength: DEFAULT_MAX_CHARACTERS_LONG }],
+
+  schema: [
+    {
+      type: 'array',
+      items: [
+        {
+          properties: {
+            maxLength: {
+              type: 'string'
```
Comment (fvictorio) [inline lib/rules/best-practises/reason-string.js]: Shouldn't `maxLength` be of type number?

File: lib/rules/best-practises/reason-string.js [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+        if (lastParameterWithoutQuotes.length > this.maxCharactersLong) {
+          this._errorMessageIsTooLong(ctx, functionName)
+        }
+      }
+    }
+  }
+
+  isReasonStringStatement(ctx) {
+    const name = this.getFunctionName(ctx)
+    return _.includes(['revert', 'require'], name)
```
Comment (fvictorio) [inline lib/rules/best-practises/reason-string.js]: PP: `return name === 'revert' || name === 'require'`.

File: lib/rules/best-practises/reason-string.js [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+  }
+
+  getFunctionName(ctx) {
+    const parent = ctx.parentCtx
+    const { 0: name } = parent.children
```
Comment (fvictorio) [inline lib/rules/best-practises/reason-string.js]: I guess this is the same as `const name = parent.children[0]`? If that's the case, please let's use that, it's much less clever.

File: lib/rules/best-practises/reason-string.js [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+    return name.getText()
+  }
+
+  getFunctionParameters(ctx) {
+    const parent = ctx.parentCtx
+    const { 2: nodes } = parent.children
```
Comment (fvictorio) [inline lib/rules/best-practises/reason-string.js]: Same comment as before.

File: lib/rules/best-practises/reason-string.js [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+    const childs = nodes.children[0].children
```
Comment (fvictorio) [inline lib/rules/best-practises/reason-string.js]: `const children`

File: lib/rules/best-practises/reason-string.js [hunk 8]
```diff
@@ -0,0 +1,101 @@
...
+  }
+
+  enterFunctionCallArguments(ctx) {
+    if (this.isReasonStringStatement(ctx)) {
+      const functionParameters = this.getFunctionParameters(ctx)
+      const functionName = this.getFunctionName(ctx)
+
+      const lastFunctionParameter = _.last(functionParameters)
+
+      const haveReasonMessage = this.haveReasonMessage(lastFunctionParameter)
```
Comment (fvictorio) [inline lib/rules/best-practises/reason-string.js]: Use `has` instead of `have`.

File: lib/rules/best-practises/reason-string.js [hunk 9]
```diff
@@ -0,0 +1,101 @@
...
+      const haveNoMessage = functionParameters.length <= 1 || !haveReasonMessage
```
Comment (fvictorio) [inline lib/rules/best-practises/reason-string.js]: The logic here is a little hard to follow. I think it can be improved by doing something like this:

```
if onlyOneParameter:
  raise noReasonMessage
  return
if lastParameterTooLong
  raise reasonTooLong
```

This won't check that the last parameter is a string, but that's probably a compiler's responsibility.

File: docs/rules.md [hunk 10]
```diff
@@ -85,3 +85,4 @@ http://solidity.readthedocs.io/en/develop/style-guide.html)
  | **function-max-lines**        | Function body contains "count" lines but allowed no more than *maxlines*. | [*\<[default](#options)\>*,&nbsp;*\<maxlines\>*] Default *maxlines* is **45**. |
  | **code-complexity**           | Function has cyclomatic complexity "current" but allowed no more than *maxcompl*. | [*\<[default](#options)\>*,&nbsp;*\<maxcompl\>*] Default *maxcompl* is **7**. |
  | **max-states-count**          | Contract has "some count" states declarations but allowed no more than *maxstates* | [*\<[default](#options)\>*,&nbsp;*\<maxstates\>*] Default *maxstates* is **15**. |
+ | **reason-string**             | Ensure that revert and require statements contains error message but allowed no more than *maxLength* | [*\<[default](#options)\>*,&nbsp;*{ maxLength: 50}*] Default *maxLength* is **20**. |
```
Comment (sohkai) [inline docs/rules.md]: Was there a particular reason why the default is set to 20? Solidity does the string allocation in 32-byte chunks.

Comment (fvictorio) [reply docs/rules.md]: Makes sense, let's use 32.

File: test/rules/best-practises/reason-string.js (hunk 1 above)
Comment (mariano-aguero) [reply test/rules/best-practises/reason-string.js]: Done!

File: lib/rules/best-practises/reason-string.js (hunk 3 above)
Comment (mariano-aguero) [reply lib/rules/best-practises/reason-string.js]: Mmm I think should be integer, change Done!

File: lib/rules/best-practises/reason-string.js (hunk 4 above)
Comment (mariano-aguero) [reply lib/rules/best-practises/reason-string.js]: Done

File: lib/rules/best-practises/reason-string.js (hunk 5 above)
Comment (mariano-aguero) [reply lib/rules/best-practises/reason-string.js]: Done

File: lib/rules/best-practises/reason-string.js (hunk 7 above)
Comment (mariano-aguero) [reply lib/rules/best-practises/reason-string.js]: Done

File: lib/rules/best-practises/reason-string.js (hunk 6 above)
Comment (mariano-aguero) [reply lib/rules/best-practises/reason-string.js]: Done

File: lib/rules/best-practises/reason-string.js (hunk 8 above)
Comment (mariano-aguero) [reply lib/rules/best-practises/reason-string.js]: Done

File: docs/rules.md (hunk 10 above)
Comment (mariano-aguero) [reply docs/rules.md]: The default value was updated to 32

File: lib/rules/best-practises/reason-string.js (hunk 9 above)
Comment (mariano-aguero) [reply lib/rules/best-practises/reason-string.js]: Logic updated

File: test/rules/best-practises/reason-string.js (hunk 2 above)
Comment (mariano-aguero) [reply test/rules/best-practises/reason-string.js]: There is no need for the last two scenarios, I remove them

**Comment by pcowgill** [comment]
@fvictorio Is this good to merge? Thanks!
